ENH: add an example with dask-ml's Incremental

stsievert opened the pull request:
I'd appreciate a review of this. I tried to point out the features and limitations of `Incremental`, and walked through a basic training + testing example. I included a note about https://github.com/dask/dask-ml/pull/247, and how Incremental is responsible for data management and sklearn is responsible for the computation on the data.

A rendered notebook is available at https://gist.github.com/stsievert/fbd2da97924e827bdb2f015aff7fe9b1


cc @TomAugspurger @mrocklin

TomAugspurger commented on machine-learning-scikit-learn.ipynb:
```diff
@@ -0,0 +1,333 @@
+    "from distributed.metrics import time\n",
+    "data = []\n",
+    "start = time()\n",
+    "for iteration in range(10):\n",
```
It would also be good to note (somewhere) that these multiple epochs + distributed provide an advantage over writing your own for loop to load in chunks of data off disk: With distributed you only pay the IO cost once.

TomAugspurger commented on machine-learning-scikit-learn.ipynb:
```diff
@@ -0,0 +1,333 @@
+    "df = pd.DataFrame(data)\n",
+    "df.plot(x='epochs', y='score')\n",
+    "plt.title('Score over iterations')\n",
+    "plt.ylabel('Score')\n",
```
"Trainining Score"

TomAugspurger commented on machine-learning-scikit-learn.ipynb:
```diff
@@ -0,0 +1,333 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "y_hat = inc.predict(X_test)\n",
```
I think scikit-learn uses `y_pred`, so people may be more familiar with that (though I'm partial to `y_hat`, in this case it's best to go where the people are).

TomAugspurger commented on machine-learning-scikit-learn.ipynb:
```diff
@@ -0,0 +1,333 @@
+   "outputs": [],
+   "source": [
+    "y_hat = inc.predict(X_test)\n",
+    "errors = np.abs(y_hat - y_test) / 2   # divide by 2 to get 0-1 labels\n",
```
`dask_ml.metrics.accuracy_score`

stsievert replied:
I was using raw dask arrays to provide some basics on how to use them. I've tried to clarify that, and re-ordered too.

TomAugspurger commented on machine-learning-scikit-learn.ipynb:
```diff
@@ -0,0 +1,333 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This accuracy might be the best possible, but there a lot of parameters to feed `SGDClassifier` that can drastically change the performance. We'll consider two of the most basic parameters, specifying a different loss function and different amount of penalty. This is the problem of hyperparameter optimization, something Dask-ML supports: https://dask-ml.readthedocs.io/en/latest/hyper-paramet ...
```
Is the first phrase missing a "not"? Somethying like

"This accuracy might not be the best possible; there are a lot of parameters..."

TomAugspurger commented on machine-learning-scikit-learn.ipynb:
```diff
@@ -0,0 +1,333 @@
+   "outputs": [],
+   "source": [
+    "from dask_ml.model_selection import GridSearchCV\n",
+    "params = {'alpha': np.logspace(-6, 0, num=10),\n",
```
I would write this like:

```python
# small parameter grid for the example
params = {
    'alpha': [0.1, 10],
    'loss': ['hingle', 'log'],
}

# An example of a larger parameter grid.
# params = {
#    'alpha': np.logspace(-6, 0, num=10),
#    'loss': ['hinge', 'log', 'modified_huber', 'squared_hinge', 'perceptron']
# }
```

TomAugspurger commented on machine-learning-scikit-learn.ipynb:
```diff
@@ -0,0 +1,333 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import seaborn as sns\n",
```
I slightly prefer 

```python
g = sns.factorplot(x='param_alpha', y='mean_test_score', data=df,
                   aspect=3,
                   hue='param_loss')
```

for visualizing these.

![gh](https://user-images.githubusercontent.com/1312546/42125075-4307149c-7c35-11e8-9b8b-29e576f73d78.png)


ticklabels would need to be cleaned up.

will also need to ensure seaborn is available on the binder worker.

stsievert replied:
I greatly prefer this instead of the heatmap.

stsievert commented on machine-learning-scikit-learn.ipynb:
```diff
@@ -0,0 +1,333 @@
+    "import numpy as np\n",
+    "n, d = int(100e3), 100\n",
+    "\n",
+    "X = da.random.normal(size=(n, d), chunks=n // 10)\n",
```
> dask/dask-ml#255 would cover this right?

It would help. I've gotten rid of `w_star` – I think I was using it for ground truth, but I don't actually use it later.

stsievert commented on machine-learning-scikit-learn.ipynb:
```diff
@@ -0,0 +1,333 @@
+   "outputs": [],
+   "source": [
+    "%%time\n",
+    "# takes about 1min\n",
```
I've reduced the number of iterations so it takes 1min with 2 workers on my machine.

TomAugspurger commented:
Hmm I had a review pending, but apparently lost it.

Overall, the content here is solid. We may want to wait on the resolution to https://github.com/dask/dask-ml/pull/258, as that'll be an API breaking change.

stsievert commented:
Could I have another review on this PR? I'm satisfied with the prose and the performance (which has spawned a couple other issues and PRs)

stsievert commented:
I've pushed the notebook with outputs. I can strip them before removal depending on #16.

TomAugspurger commented on machine-learning-scikit-learn.ipynb:
```diff
@@ -0,0 +1,744 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Scikit-learn is a popular machine learning library whose API is easy to use because it has been inspired by other machine learning libraries. It encompasses a wide variety of problem formulations, ranging from simple linear models to more complex clustering or decomposition formulations. Scikit-learn works great for problems that fit on a single machine, but struggles for larger problems. Th ...
```
I'm not sure about the origins of scikit-learn's design, but I often see *it* cited as the inspiration for other APIs (e.g. https://spark.apache.org/docs/latest/ml-pipeline.html#main-concepts-in-pipelines)

mrocklin commented:
Some comments, mostly language related.  My apologies for not going in-line:

-  Avoid words like *plethora* that may not be known to non-native english speakers
-  Similarly with phrases like "ever hear of teaching to the test?"  This can be confusing if you don't know what it means, which is possibly most of our readers.  This then diminishes leads you into a tangent of trying to figure out if teaching is another term for training/fitting, etc..
- More generally, we should probably avoid jokes and cleverness in all prose except perhaps blogposts.  If you were to reduce the amount of prose in this notebook by a factor of two or three I think that that would be for the best.  A user probably wants to move through this in a couple minutes.  We should endeavor to give them the minimum amount of information that they need to move on.  They will appreciate that we respect their time.
-  Remove phrases like "as usual" and "of course" which assume that the user knows what's going on.  A wide variety of users will read these examples, and for many of them all of this will be new.
-  I recommend removing the manual computation of accuracy.  I get that it can be useful to show them dask.arrays, but they can also look at the dask array examples for this if that's their interest.  Lets focus in on the use of Incremental
-  I also recommend removing the GridSearch section for the same reason.  Lets get them in and out quickly.  
- Calling `foo.persist(); bar.persist()` is sort of an anti-pattern.  It's nicer to call them all at once with the dask.persist function.  

As an example, here is what I recommend for how to handle persist and da.unique.  I recommend the following cells directly after creating the training data.

```
### Persist data in memory

This dataset is small enough to fit in distributed memory, so we call `dask.persist` now.

    X_train, y_train, X_test, y_test = dask.persist(X_train, y_train, X_test, y_test)

If you are working in a situation where your dataset does not fit in memory then you should skip this step.  Everything will still work, but will be slower and use less memory.

### Pre-compute classes

We pre-compute the classes from our training data

    classes = da.unique(y_train).compute()
    classes

...

### Create Scikit-Learn Estimator
 
Your explanation looks good (other than "as usual")

    sgd = ...

### Wrap with Dask-ML Incremental 

Incremental is a meta-estimator that ...
```

I've found that this level of prose/code is easily digestible by peole with short attention spans (most readers).  Separating things out by section headers with brief informative names helps them to understand the scope of the step of the computation that they are about to learn about.

TomAugspurger commented on machine-learning-scikit-learn.ipynb:
```diff
@@ -0,0 +1,744 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Creating a distributed scheduler will provide good feedback because we can view the dask.distributed dashboard. This will provide progress and performance metrics, including visualization of jobs that are running."
```
I think this is redundant with the previous.

TomAugspurger commented on machine-learning-scikit-learn.ipynb:
```diff
@@ -0,0 +1,744 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now, let's call `partial_fit` multiple times.\n",
```
I would remove this block, since implementing `max_iter` for Incremental will I think make it obsolete?

stsievert replied:
I want to leave this block we're show casing `Incremental`, which will very likely be used for other tools to get some idea or progress/etc. I would imagine that the `data` dictionary I use will be used, and probably with more data.

I have restructered a bit, to move the persisting data section below the "calling `partial_fit`" header.

TomAugspurger commented on machine-learning-scikit-learn.ipynb:
```diff
@@ -0,0 +1,744 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "There a lot of parameters to feed `SGDClassifier`, and this accuracy depends on those parameters, and it may not be as high as it can be. We'll consider two of the most basic parameters, specifying a different loss function and different amount of penalty. This is the problem of hyperparameter optimization, something Dask-ML supports: https://dask-ml.readthedocs.io/en/latest/hyper-parameter- ...
```
Typo "There a".

stsievert commented:
I have addressed both @mrocklin and @TomAugspurger's comments. I've pushed the notebook with the outputs again.

> I also recommend removing the GridSearch section for the same reason. Lets get them in and out quickly.

Good idea. I've removed the model selection section. I have briefly mentioned this alongside the Dask joblib integration in a "Learn more" section.

TomAugspurger commented:
FYI I’m traveling today and will likely be busy on pandas things the next few days, so don’t let me hold any of these up.

________________________________
From: Scott Sievert <notifications@github.com>
Sent: Wednesday, July 4, 2018 10:12:25 PM
To: dask/dask-examples
Cc: Tom Augspurger; Mention
Subject: Re: [dask/dask-examples] ENH: add an example with dask-ml's Incremental (#15)


I have addressed both @mrocklin<https://github.com/mrocklin> and @TomAugspurger<https://github.com/TomAugspurger>'s comments. I've pushed the notebook with the outputs again.

I also recommend removing the GridSearch section for the same reason. Lets get them in and out quickly.

Good idea. I've removed the model selection section. I have briefly mentioned this alongside the Dask joblib integration in a "Learn more" section.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/dask/dask-examples/pull/15#issuecomment-402595044>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABQHItrOMEkjqJIP-9JWVitHCwL9Yasdks5uDYQZgaJpZM4U9vR5>.

mrocklin commented:
I'm taking a look through this now. I may make a few edits directly if that's ok.

stsievert commented:
> I may make a few edits directly if that's ok.

Thank you, it's appreciated. That's more than okay.

stsievert commented on machine-learning/incremental.ipynb:
````diff
@@ -0,0 +1,542 @@
+    "    df = pandas.read_csv(filename)\n",
+    "    x, y = ...  # get data from CSV using the Pandas API\n",
+    "    est.partial_fit(x, y)\n",
+    "```\n",
````
Do you have any examples for "this is a common operation"? The first thing I would do is shuffle the dataset before calling the optimizer on each block.

mrocklin replied:
> The biggest edits I made were adding some motivation for calling dask.persist.

Sounds fine to me.  As a general note, there are always going to be lots of things that we'll do in any of these examples that are important that we won't be able to tell them about.  We'll necessarily skip a lot of potential wisdom we could deliver in order to focus on the topic at hand.

In this case though I agree that including some information about persist is good, especially because it's nice and concise.

mrocklin commented:
This is in.  Thanks @stsievert !
